pgmpy/base/MAG.py

```diff
@@ -0,0 +1,295 @@
+        True
+        """
+
+        is_inducing = True
+        for path in nx.all_simple_paths(self, source=u, target=v):
+            if len(path) <= 2:
+                continue
+
+            for i in range(1, len(path) - 1):
+                prev_node, curr_node, next_node = path[i - 1], path[i], path[i + 1]
+
+                if not self._is_collider(prev_node, curr_node, next_node):
+                    is_inducing = False
+                    break
+
+                ancestors_uv_vu = self.get_ancestors(u).union(self.get_ancestors(v))
+                if curr_node not in W and curr_node not in ancestors_uv_vu:
+                    is_inducing = False
+                    break
+
+        return is_inducing
+
+    def is_visible_edge(self, u, v) -> bool:
+        """
+        Check if an edge is visible.
+
+        An edge is visible if it exists and is not shielded by an inducing path
+        through latent variables.
+
+        Parameters
+        ----------
+        u : Hashable
+            First node.
+
+        v : Hashable
+            Second node.
+
+        Returns
+        -------
+        bool
+            True if the edge is visible, False otherwise.
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag = MAG()
+        >>> mag.add_edge("X", "Y", "-", ">")
+        >>> mag.is_visible_edge("X", "Y")
+        True
+        """
+        if not self.has_edge(u, v):
+            return False
+
+        graph_without_edge = self.copy()
+        graph_without_edge.remove_edge(u, v)
+        return not graph_without_edge.has_inducing_path(u, v, self.latents)
+
+    def lower_manipulation(self, X):
+        """
+        Perform lower manipulation (marginalization).
+
+        Removes variables in `X` from the MAG while preserving independence
+        structure implied by marginalization. Invisible edges are replaced with
+        bidirected edges. Directed edges into marginalized nodes are removed.
+
+        Parameters
+        ----------
+        X : set
+            Set of nodes to marginalize (remove).
+
+        Returns
+        -------
+        MAG
+            A new MAG with nodes in X marginalized out.
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag = MAG()
+        >>> mag.add_edge("X", "L", "-", ">")
+        >>> mag.add_edge("Y", "L", "-", ">")
+        >>> mag.latents = {"L"}
+        >>> new_mag = mag.lower_manipulation({"L"})
+        >>> list(new_mag.edges())
+        [('X', 'Y')]
+        """
+        new_mag = self.copy()
+
+        for node in X:
+            neighbors = list(self.neighbors(node))
+
+            for i in range(len(neighbors)):
+                for j in range(i + 1, len(neighbors)):
+                    u, v = neighbors[i], neighbors[j]
+
+                    if self._is_collider(u, node, v):
+                        new_mag.add_edge(u, v, ">", ">")
+
+        new_mag.remove_nodes_from(X)
```

Comment:
I don't think any of these manipulations is supposed to remove the node from the graph. Can you double-check?